### ISSUE TITLE
apidoc doesn't look for package.json in current working directory

### ISSUE THREAD

**Issue by jeremybradbury** [apidoc/apidoc#1086]
# Describe the bug
Neither the config file, nor the package.json rules are being respected. Both are ignored, removing base urls & example, leaving default titles.

# Steps to reproduce

Follow the documentation like I did?

1. - `npm install apidoc -g`
2. - add comments
3. - `apidoc -i src -o apidoc`
4. - witness failure

# Information

- Version of apidoc: v0.50.1
- Operating system (e.g. GNU+Linux/MacOS/Windows): Windows 10 (v11 ready, invite waiting... probably a while)
- Node version: 14.17.0
- Browser (e.g. firefox, chrome): Brave/Chrome/Edge (shouldn't matter for reading static content)

**Comment by NicolasCARPi** [comment]
Can you run it with `-d` flag and post output please? Also, you need to specify the path to config file with `-c` if for some reason it cannot find it by himself. The debug will tell us.

**Comment by jeremybradbury** [comment]
Thanks for the quick reply! Took GitHub probably 15 minutes to fire off the email.

```apidoc -i .\app -o .\docs -d
[debug] Parsed options:

{
  excludeFilters: [ 'apidoc.config.js', 'node_modules' ],
  includeFilters: [
    '.*\\.(clj|cls|coffee|cpp|cs|dart|erl|exs?|go|groovy|ino?|java|js|jsx|kt|litcoffee|lua|mjs|p|php?|pl|pm|py|rb|scala|ts|vue)$'
  ],
  src: [ '.\\app' ],
  dest: '.\\docs',
  mode: '',
  template: 'C:\\Users\\info\\AppData\\Roaming\\nvm\\v14.17.0\\node_modules\\apidoc\\template\\',
  config: '',
  apiprivate: false,
  verbose: false,
  single: false,
  debug: true,
  colorize: true,
  filters: {},
  languages: {},
  parsers: {},
  workers: {},
  silent: false,
  dryRun: false,
  markdown: true,
  lineEnding: '\r\n',
  encoding: 'utf8',
  copyDefinitions: true,
  filterBy: '',
  logFormat: 'simple',
  warnError: false
}
debug: Now looking for apidoc config files
debug: Now looking for package.json
debug: Now looking for apidoc.json
debug: Now looking for apidoc.config.js
verbose: apidoc-generator name: apidoc
verbose: apidoc-generator version: 0.50.1
verbose: apidoc version: 0.50.1
verbose: apidoc-spec version: 0.3.0
debug: No plugins found.
debug: load parser language: .clj, ./languages/clj.js
debug: load parser language: .coffee, ./languages/coffee.js
debug: load parser language: .erl, ./languages/erl.js
debug: load parser language: .ex, ./languages/ex.js
debug: load parser language: .exs, ./languages/ex.js
debug: load parser language: .litcoffee, ./languages/coffee.js
debug: load parser language: .lua, ./languages/lua.js
debug: load parser language: .pl, ./languages/pm.js
debug: load parser language: .pm, ./languages/pm.js
debug: load parser language: .py, ./languages/py.js
debug: load parser language: .rb, ./languages/rb.js
debug: load parser language: default, ./languages/default.js```

**Comment by NicolasCARPi** [comment]
> debug: Now looking for package.json

I think it simply doesn't find the package.json file. Is it in `app/`? It doesn't look for it in parent folder of `src`/`input`.

**Comment by jeremybradbury** [comment]
@NicolasCARPi if that's the case, this def is a bug.

I ran this at the project root which contains an app folder, the only folder I wanted to scan. 

You shouldn't need the package.json to exist inside the input folder, it should look in the cwd where the command was ran.

**Comment by NicolasCARPi** [comment]
>  it should look in the cwd where the command was ran.

Yep, agreed. I'll add this to the code.

**Comment by NicolasCARPi** [comment]
What do you think is best?

1. Scan cwd first and then the src, if a config file is found in src it will supersede any other
2. merge the configs from files found. So a package.json might define keys that are not defined in app/apidoc.json, and app/apidoc.json might override keys from package.json.

Right now the code is doing 1 (with the change to scan for cwd of course) but I think 2 is better. But 1 also makes sense IMHO.

**Comment by jeremybradbury** [comment]
I agree, #2. 

cwd/*.json = the lowest level, nothing should be more general than this folder
-i folder/*.json = override anything in the parent
-c filename.json = override everything else

**Comment by NicolasCARPi** [comment]
can you try the dev branch please and let me know if it looks alright?

**Comment by jeremybradbury** [comment]
Hey sorry I wasn't able to get to this over the weekend. I did a quick code review on the commit and it looks good.

I will have time tomorrow (Tuesday) evening to give it a spin from the repo. 

Thanks so much for even asking for my help with this & getting it tested in context of the original issue. This is a great example of how to handle issues well.

And by the way my employers are loving the tool so far.

**Comment by jeremybradbury** [comment]
It worked beautifully and behaved as expected. It found my pacakge.json which I removed the object from, it found my apidoc.json file (output below with trimmed pathnames). 

```
$ node repos\apidoc\bin\apidoc -i app -o docs
Done in 11.38s.
$ node repos\apidoc\bin\apidoc -i app -o docs --debug
[debug] Parsed options:
{
  excludeFilters: [ 'apidoc.config.js', 'node_modules' ],
  includeFilters: [
    '.*\\.(clj|cls|coffee|cpp|cs|dart|erl|exs?|go|groovy|ino?|java|js|jsx|kt|litcoffee|lua|mjs|p|php?|pl|pm|py|rb|scala|ts|vue)$'
  ],
  src: [ 'app' ],
  dest: 'docs',
  template: 'D:\\Docs\\repos\\apidoc\\template\\',
  config: '',
  apiprivate: false,
  verbose: false,
  single: false,
  debug: true,
  colorize: true,
  filters: {},
  languages: {},
  parsers: {},
  workers: {},
  silent: false,
  dryRun: false,
  markdown: true,
  lineEnding: '\r\n',
  encoding: 'utf8',
  copyDefinitions: true,
  filterBy: '',
  logFormat: 'simple',
  warnError: false
}
debug: Now looking for apidoc config files
debug: Now looking for package.json
debug: Found file: repos\*\package.json 
debug: Now looking for apidoc.json
debug: Found file: repos\*\apidoc.json
debug: Now looking for apidoc.config.js
verbose: apidoc-generator name: apidoc
verbose: apidoc-generator version: 0.50.1
verbose: apidoc version: 0.50.1
verbose: apidoc-spec version: 0.3.0
debug: No plugins found.
```

When I deleted the apidoc.json file it seemed to fallback to the default settings, as expected.

It also worked moving the config object back to package.json.

```
debug: Now looking for apidoc config files
debug: Now looking for package.json
debug: Found file: C:\Users\info\repos\mos-oauth\package.json
debug: Using apidoc key of package.json
```

The one thing of note is that last line only showed when using package.json and not when using apidoc.json. 

I personally don't need that, but I had a thought on it.

It may warrant being an "info" line or 2 to stdout, even without the debug flag, something show which config file was used, perhaps add a warning if the default config object is used because no file was found. Something easy to ignore when you don't care, but a useful reminder when you do.

Thanks for all your time & super quick responses. If you want to close this issue, I am very satisfied with the resolution as is. 

I'll leave the closing it part in your court, as you may want to release the change, before closing so it stays in your queue.

**Comment by NicolasCARPi** [comment]
> The one thing of note is that last line only showed when using package.json and not when using apidoc.json.

Yes because you can find a package.json without an apidoc key, so it only tells you about it if there is one. I've set it to `verbose` instead of `debug`.

> add a warning if the default config object is used because no file was found

There was a bug in this check so the condition was never true, I fixed that so now a warning message will be shown if no config file is found.

> I'll leave the closing it part in your court, as you may want to release the change, before closing so it stays in your queue

It'll be closed automatically when the commit is merged to master, because the commit message has a "fix #1086" in it ;) And yes it's better to leave the issue open until the fix is merged and published.
